documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,555 @@
+# Diagnostic IPC Protocol
+
+## Overview
+
+This spec describes the IPC Protocol to be used for communicating with the dotnet core runtime's Diagnostics Server from an external client over a platform-specific transport, e.g., Unix Domain Sockets.
+
+
+### Terminology
+
+The protocol will use the following names for various constructs and behaviors defined in this spec:
+* *Diagnostic IPC Protocol*: The protocol defined in this spec
+* *Diagnostic Server*: The server in the runtime that receives/sends Diagnostic IPC Procotol communication.
+* *Commands*: The functionality being invoked in the runtime that communicates over the Diagnostic IPC Protocol, e.g., "Start an EventPipe stream".  These are encoded as a `command_set` and a `command_id`.
+* *Flow*: A sequence of interactions making up communication with the Diagnostics Server
+* *Pipe*: The duplex communication construct this protocol is communicated over.  This is a Unix Domain Socket on *nix systems and a Named Pipe on Windows.
+* *IPC Message*: The base unit of communication over the Diagnostic IPC Protocol. Is made up of a Header and a Payload.
+* *Header*: A struct containing version and metadata.
+* *Payload*: An opaque chunk of data that is Command specific.
+* *Optional Continuation*: The reuse of the pipe for application specific communication. This communication does not need to adhere to any requirements listed in this spec, e.g., this could be a stream of custom encoded data that is Command specific.
+
+### General Flow
+
+All communication with the Diagnostic Server will begin with a Diagnostic IPC Message sent from the client to the server.  The server will respond with a Diagnostic IPC Message.  After this, the client and runtime _may_ reuse the Pipe for any Command specific communication which is referred to as an Optional Continuation.
+
+![Generic Flow](ipc-protocol-genericflow.svg)
+
+```
+runtime <- client : [ Header ][ Payload ]
+runtime -> client : [ Header ][ Payload ]
+optional:
+    runtime <-> client : [ Optional Continuation ]
+connection closed
+```
+
+Example flow for EventPipe:
+```
+runtime <- client : [ ver; size; EventPipe CollectTracing ][ stream config struct  ] <- Diagnostic IPC Message
+runtime -> client : [ ver; size; EventPipe TracingStarted ][ stream started struct ] <- Diagnostic IPC Message
```

Comment:
There isn't actually a specific response for TracingStarted right? It might be better to say 'OK'